\documentclass{article} % For LaTeX2e
\usepackage{nips15submit_e,times}
\usepackage{hyperref}
\usepackage{url}
\usepackage{natbib}
\usepackage{graphicx}

\title{PeakSegJoint: supervised detection of the same peaks jointly
  across several ChIP-seq samples}


\author{
David S.~Hippocampus\thanks{ Use footnote for providing further information
about author (webpage, alternative address)---\emph{not} for acknowledging
funding agencies.} \\
Department of Computer Science\\
Cranberry-Lemon University\\
Pittsburgh, PA 15213 \\
\texttt{hippo@cs.cranberry-lemon.edu} \\
\And
Coauthor \\
Affiliation \\
Address \\
\texttt{email} \\
\AND
Coauthor \\
Affiliation \\
Address \\
\texttt{email} \\
\And
Coauthor \\
Affiliation \\
Address \\
\texttt{email} \\
\And
Coauthor \\
Affiliation \\
Address \\
\texttt{email} \\
(if needed)\\
}

% The \author macro works with any number of authors. There are two commands
% used to separate the names and addresses of multiple authors: \And and \AND.
%
% Using \And between authors leaves it to \LaTeX{} to determine where to break
% the lines. Using \AND forces a linebreak at that point. So, if \LaTeX{}
% puts 3 of 4 authors names on the first line, and the last on the second
% line, try using \AND instead of \And before the third author name.

\newcommand{\fix}{\marginpar{FIX}}
\newcommand{\new}{\marginpar{NEW}}

%\nipsfinalcopy % Uncomment for camera-ready version

\begin{document}


\maketitle

\begin{abstract}
  Joint peak detection is a central problem when comparing samples in
  genomic data analysis, and current algorithms for this task are
  unsupervised and mostly effective for a single data type and pattern
  (e.g. H3K4me3 data with a sharp peak pattern). We propose
  PeakSegJoint, a new constrained multi-sample maximum likelihood
  segmentation model. To select the number of peaks in the
  segmentation, we propose a new supervised penalty learning model. To
  infer the parameters of these two models, we propose to use a
  discrete optimization heuristic for the segmentation, and convex
  optimization for the penalty learning. We show that this method
  achieves state-of-the-art peak detection, and results in
  interpretable peaks that occur in exactly the same positions across
  samples.
\end{abstract}

\section{Introduction}

\subsection{Peak detection in ChIP-seq data}

Chromatin immunoprecipitation sequencing (ChIP-seq) is a biological
experiment for genome-wide profiling of histone modifications and
transcription factor binding sites, with many experimental and
computational steps \citep{practical}. Briefly, each experiment yields
a set of sequence reads which are aligned to a reference genome, and
then the number of aligned reads are counted at each genomic position
(Figure~\ref{fig:good-bad}). These data can be interpreted using
one of the many available peak detection algorithms
\citep{evaluation2010, rye2010manually, chip-seq-bench}, which each
essentially work as a binary classifier for each genomic position. The
positive class is enriched (peaks) and the negative class is
background noise. Importantly, peaks and background occur in long
contiguous segments across the genome.

\section{Related work}

\citet{hierarchical-joint} describe an HMM. TODO describe.

\citet{JAMM} describe JAMM a joint method for analyzing replicates. It
says that it and DFilter are both ``universal'' for broad and narrow
peaks. They used TF bindings sites as positive controls. They say that
it is better to analyze replicates independantly than to pool
replicates.

\citet{PePr} describe PePr a method for identifying consistent or
differential peaks across replicates. They applied it to TF data sets,
and used visual inspection to test (but not train).

\begin{figure}[b!]
  \centering
  \includegraphics[width=\textwidth]{figure-PeakSegJoint}
  \caption{Blue line segments show the sequence of PeakSegJoint models
    for $S=3$ samples and $p\in\{0, 1, 2, 3\}$ peaks.}
  \label{fig:PeakSegJoint}
\end{figure}

\section{Model}

\section{Algorithm}

\section{Results}

\section{Discussion}

\section{Conclusions}

\bibliographystyle{abbrvnat}
\bibliography{refs}

\end{document}
